src/censys/cloud_connectors/azure_connector/connector.py

```diff
@@ -56,72 +71,148 @@ def __init__(self, settings: Settings):
         """Scan all Azure Subscriptions."""
         provider_settings: dict[
             tuple, AzureSpecificSettings
         ] = self.settings.providers.get(self.provider, {})
+
+        self.logger.debug(
+            f"scanning {self.provider} using {self.settings.scan_concurrency} processes"
+        )
+
+        label_prefix = self.get_provider_label_prefix()
+        pool = Pool(processes=self.settings.scan_concurrency)
+
         for provider_setting in provider_settings.values():
+            # this is so confusing - plural settings to setting?
             self.provider_settings = provider_setting
-            self.credentials = ClientSecretCredential(
-                tenant_id=provider_setting.tenant_id,
-                client_id=provider_setting.client_id,
-                client_secret=provider_setting.client_secret,
-            )
+
+            # self.credentials = ClientSecretCredential(
+            # credentials = ClientSecretCredential(
+            #     tenant_id=provider_setting.tenant_id,
+            #     client_id=provider_setting.client_id,
+            #     client_secret=provider_setting.client_secret,
+            # )
+
             for subscription_id in self.provider_settings.subscription_id:
                 self.logger.info(f"Scanning Azure Subscription {subscription_id}")
-                self.subscription_id = subscription_id
+                # self.subscription_id = subscription_id
+
                 try:
+                    possible_labels: set[str]
                     if self.scan_all_regions:
-                        self.get_all_labels()
+                        possible_labels = self.get_all_labels(
+                            subscription_id, label_prefix
+                        )
+                    else:
+                        possible_labels = set()
+
+                    scan_context = AzureScanContext(
+                        provider_settings=provider_setting,
+                        subscription_id=subscription_id,
+                        credentials=None,
+                        # credentials=credentials,
+                        possible_labels=possible_labels,  # self.possible_labels,
+                        scan_all_regions=self.scan_all_regions,
+                        label_prefix=label_prefix,
+                    )
 
-                    self.scan()
+                    # self.scan(**{"scan_context": scan_context})
+                    # self.scan_seeds(**{"scan_context": scan_context})
+                    # pool.apply_async(
+                    pool.apply_async(
+                        # self.scan_seeds,
+                        self.scan,
+                        kwds={
+                            "scan_context": scan_context,
+                        },
+                        error_callback=lambda e: self.logger.error(f"Async Error: {e}"),
+                    )
 
+                    # TODO: figure out how to make this wait until scan is finished:
```

Comment:
** After writing this, # 2 is definitely better and is a fine solution. Figured I would leave my thought process though **
- - -

I don't think this solution is compatible with multiprocessing. Since we don't scan by region, but region is part of the label, options I see are:
1. setting a ttl on the asset and only submitting what we find
   Pros: easy
   Cons: could leave seeds "stale" for up to 48 hours (it's probably ~24 hours now depending on attribution)
2. One label will correspond to one process in this case, so we should be able to move the submission of empty payloads to labels_not_found within the single process. In this case it would be right after `super().scan(**kwargs)` in `self.scan()`. If we split up seeds/assets like we talked about for aws/gcp:
  ```
scan_all
  scan_seeds()
    super.scan_seeds()
      get_seeds()
  scan_cloud_assets()
    super.scan_cloud_assets()
      get_cloud_assets()
```
then it would would be at the end of `scan_seeds()` after it calls `super.scan_seeds()`, where "it" is
```
if self.scan_all_regions:
  for label_not_found in scan_context.possible_labels:
    self.delete_seeds_by_label(label_not_found)
```

Reply:
Let's talk about this one some more.

My concern is that we will further split up how the connector works over time. I don't know that we can rely on an account or any level of region being in a single process over time. I could see each resource type being spawned off as well. 

Another strategy could be to keep track of the "run id" or create a single value for the entire run. Upon completion, we could instruct items that aren't part of the current run should be removed. 

The cloud provider change streams might be a better way to handle this prune routine as well.

It will be hard to manage our concurrency patterns with features like this over time. With large enough customers we might not be able to hold all of the possible labels in ram either. I frequently saw Out Of Memory (OOM) errors due to patterns like this while trying to build ingestion pipelines. Those instances had something like 16 GB of ram.